### ISSUE TITLE
ARM64 Support for VerneMQ

### ISSUE THREAD

**Issue by drf** [erlio/vernemq#461]
Putting here what we have found out on Slack about getting VMQ up and running on ARM64. The main target here is not the RPi-like boards, rather new ARM64 servers such as Scaleway's ones (https://www.scaleway.com/armv8-cloud-servers/).

These are the actions required (up to now):

- [ ] Get leveldb to compile and run on ARM64
  - [x] Fork leveldb into erlio
  - [x] Import ARM64 patch into new fork (PR available: https://github.com/erlio/leveldb/pull/1)
  - [ ] Release a new leveldb which can be used by eleveldb
- [ ] Get eleveldb to compile and run on ARM64
  - [ ] Have eleveldb use the new erlio-specific release of leveldb with the needed patch
  - [ ] Update snappy or patch the buildsystem (autoconf) to support ARM64
- [ ] Get VMQ to run on ARM64
  - [ ] Add an ARM64 target to Travis via QEMU, and make tests green.

**Comment by drf** [comment]
ofc I can take care of pretty much all of the tasks, but we need general consensus especially on (e)leveldb and on what we should do there. Paging @larshesel and @dergraf on this.

**Comment by larshesel** [comment]
What do you think is needed to get it working? Could you outline the different viable options as you seem them? We're of course interested in the simplest and smallest change which will make it work on ARM64 to minimize the possibility of introduce any adverse effects.

**Comment by ioolkos** [comment]
I think @drf's already outlined kind of the minimal set of actions needed. 
I think the main thing to discuss is whether or not we want to patch and re-release leveldb in our own fork. I'm fine with this, but maybe someone sees technical issues arise (cc @larshesel / @dergraf )

**Comment by drf** [comment]
What @ioolkos said. The least disruptive options is just to patch out autotools on Snappy's side (which would prevent any kind of changes to the code), but here the main issue is the release-fu needed, so we need a general agreement from maintainers on what needs to be done

**Comment by larshesel** [comment]
@drf did you try to upgrade the Snappy autotools? would that work as well?

**Comment by drf** [comment]
@larshesel upgrading the whole buildsystem is 100% hell, adding a small patch to configure is enough to make it work (the only issue the BS has ATM is that it doesn't understand what aarch64 means). That is pretty much a no brainer, if we decide to go that way I can give it a shot as soon as I'm back from the seaside :D

**Comment by larshesel** [comment]
ok - let's try and go with the patch.

I'd like to get @dergraf to chime in with his thoughts on this as well..

**Comment by ioolkos** [comment]
"Forward, said the engineer"! 
Still my favourite Jules Verne quote :)

**Comment by drf** [comment]
@larshesel ok, let's wait for @dergraf as I won't tackle this before a week anyway :)

**Comment by dergraf** [comment]
go for the patch! :+1:

**Comment by larshesel** [comment]
hi @drf is this still something you're interested in / working on?

**Comment by drf** [comment]
Hi @larshesel , unfortunately this has become quite low priority due to the fact we found out a number of other issues with ARM64 servers :/ however, I can still finish my work on this one, given it's not a very big deal. Remember I'm blocked by erlio/leveldb#1 though! :D Afterwards, I can start moving pieces around to make the magic happen.

On a side note, we'll soon tackle #234 with a solid proposal :)

**Comment by larshesel** [comment]
Hi @drf, I just merged erlio/leveldb#1 - but don't spend time on it unless you really want to - it's very few people who've expressed interest in this so far... And I'd still think some hours would have to be spent in making eleveldb use erlio/leveldb version of leveldb as well as getting the autoconf stuff to work and setting up QEMU on travis...

Out of curiosity, which other issues did you run into regarding ARM64 servers?

Looking forward to see the proposal for #324 :)

**Comment by drf** [comment]
@larshesel ok, then I guess we might put this on hold until there is a community interest for this... Mostly because I'm afraid of the effort of keeping everything tested and stable with a small userbase :/

In the end, Erlang on ARM64 isn't exactly the most well tested platform, the server SoCs aren't as reliable as we initially thought, and, in the end, it doesn't cut it from a production standpoint (yet). It's a shame, as the fact of having a lot of cores for a very cheap price is definitely juicy for something like Verne and for BEAM in general. But I guess we need a much bigger audience for this to take off. In fact, from what I've read, (limited) community interested is mostly for RPis&friends.

**Comment by larshesel** [comment]
ok - let's put this on hold then - I'll remove the erlio/leveldb repo as it's not used for anything else (and would just cause confusion).

Closing this for now.

**Comment by alex-savin** [comment]
Might be it is a good time to get back to arm64 support?!

**Comment by NikolayVasilyev** [comment]
There is a plan to use VerneMQ in a distributed system based on arm64 architecture.
But unfortunately it fails building `eleveldb`.

Is there any hope to reopen this issue?

Are there any suggestions on how to move on?

**Comment by ioolkos** [comment]
Hi @NikolayVasilyev have you tried building the RPI32 target? it has a couple tricks to make LevelDB build.

In principle, there's also another possible approach: switching from LevelDB to some other flavour like Leveled (for the metadata store), which is full Erlang, and switching to ETS/DETS for the message store backend.

(most of this stuff is in master, but currently undocumented)
I'd check if the RPI build helps you. It's really only about LevelDB.

**Comment by NikolayVasilyev** [comment]
Thanks for reply.

Yes, I have already made a working prototype using raspberry pi and `rpi32` build.

As I figured out from the `rebar.config` the only principal change is to use the `rpi-32` branch from `basho/eleveldb`.

The `basho/eleveldb` changes are the following:

 * using `unsigned long` instead of `size_t` in `c_src/eleveldb.cc`
 * using `lager` instead of `cuttlefish`
 * using a custom make file to build sources
 * and some mysterious changes in `rebar.config.script` which I have no idea what they are doing

Are this changes cover all `arm32` architecture specific?

Would it be enough to make a similar changes for `arm64` architecture?
